Coupling CTSM with WRF

negin513 opened the pull request:
TYPE: New Feature

KEYWORDS: WRF-CTSM coupling, Community Terrestrial Systems Model (CTSM), Community Land Model (CLM), 
coupled land-atmosphere models

SOURCE: Negin Sobhani <sup>1</sup> , Dave Lawrence <sup>1</sup> , Bill Sacks <sup>1</sup>, 
Sam Levis <sup>2</sup>, Mariana Vertenstein <sup>1</sup>, Dave Gill <sup>1</sup>, Mike Barlage  <sup>1,3</sup>, 
and Fei Chen <sup>1</sup> 

<sup>1</sup> National Center for Atmospheric Research (NCAR)
<sup>2</sup> SLevis Consulting 
<sup>3</sup> National Centers for Environmental Prediction (NCEP)

DESCRIPTION OF CHANGES:
This PR includes the capability to couple WRF with Community Terrestrial Systems Model (CTSM) via LILAC (Lightweight 
Infrastructure for Land-Atmosphere Coupling). CTSM can be used with WRF as one of WRF's land surface model 
schemes by setting `sf_surface_physics` in the WRF namelist.input file:

```
&physics
sf_surface_physics = 6, 6, 6,
/
```

For instructions on how to run WRF with CTSM please check [instructions on using CTSM with WRF.](https://escomp.github.io/ctsm-docs/versions/master/html/lilac/specific-atm-models/index.html)

LIST OF MODIFIED FILES: 
M       Registry/Registry.EM_COMMON
M       arch/Config.pl
M       arch/configure.defaults
M       arch/postamble
M       arch/preamble
A        doc/README.CTSM
M       dyn_em/module_first_rk_step_part1.F
M       dyn_em/module_initialize_real.F
M       dyn_em/solve_em.F
M       main/depend.common
M       phys/Makefile
M       phys/module_physics_init.F
A       phys/module_sf_ctsm.F
M       phys/module_surface_driver.F
M       share/input_wrf.F
M       share/module_check_a_mundo.F
M       wrftladj/solve_em_ad.F 

TESTS CONDUCTED:
1. The jenkins tests are all passing.
2. Initial assessment of the performance of WRF-CTSM5(NWP) compared to WRF-Noah and WRF-NoahMP has been 
completed over the CONUS domain.  Overall, we are finding that WRF-CTSM5(NWP) is performing reasonably well, in 
some cases even showing improved performance improvement. For example, the following plot shows the bias for 
maximum daily temperature from NOAH, NOAH-MP, and CTSM for June 2013 over METAR observation sites:

![CTSM_PR_2013-04_Tmin_Tmax](https://user-images.githubusercontent.com/17344536/90686072-05acc780-e228-11ea-91fa-4d239e548d1b.png)

RELEASE NOTE: 
The capability to couple CTSM with WRF is added and activated by setting namelist option `sf_surface_physics` to 6. This is the initial beta release of WRF-CTSM coupling capability. For instructions on how to run WRF with CTSM please check [instructions on using CTSM with WRF.](https://escomp.github.io/ctsm-docs/versions/master/html/lilac/specific-atm-models/index.html)

davegill commented:
@negin513 
Negin,
Would you include the text from the jenkins test in one of these lower comment boxes, please.

negin513 commented:
Test Type              | Expected  | Received |  Failed
    = = = = = = = = = = = = = = = = = = = = = = = =  = = = =
    Number of Tests        : 19           18
    Number of Builds       : 48           46
    Number of Simulations  : 166           164        0
    Number of Comparisons  : 105           104        0

    Failed Simulations are: 
    None
    Which comparisons are not bit-for-bit: 
    None

negin513 commented:
@davegill Did you mean the test results I posted above? Or would you like the results from the attached zip file in the Jenkins email?

davegill commented:
@negin513 
Negin,
Yes, what you put in the comments is what we wanted to see.
1. Would you add a doc/README.CTSM file. In can be short. Info on where to go for web resources, etc. A few sentences.
2. Do you have an easy example that you can show using WRF+CTSM vs WRF with any other land model? A couple of figures to demonstrate that things seem to be working OK. That can go in the "test conducted" section.

Let's check your permissions. Would you add a label that says "Feature Branch".

negin513 commented:
> Let's check your permissions. Would you add a label that says "Feature Branch".
Dave, I can change (or add) any labels to this PR. I think I don't have the permission for it.

davegill commented on arch/configure.defaults:
```diff
@@ -16,7 +16,7 @@ CC              =       CONFIGURE_CC
 RWORDSIZE       =       CONFIGURE_RWORDSIZE
 PROMOTION       =       #-fdefault-real-8
-ARCH_LOCAL      =       -DNEC -DNONSTANDARD_SYSTEM_SUBR  -DWRF_USE_CLM
+ARCH_LOCAL      =       -DNEC -DNONSTANDARD_SYSTEM_SUBR  CONFIGURE_D_CTSM
```
@negin513 
Negin,
For your `CONFIGURE_D_CTSM` flag modification, you have accidentally removed the `-DWRF_USE_CLM` cpp flag.

negin513 replied:
Dave, I agree with you. I think we should keep the `-DWRF_USE_CLM` as it is not relevant to CTSM coupling at all. 
I think @ billsacks removed this flag while working on this and I want to confirm with him if it is okay to add `-DWRF_USE_CLM` back to `arch/configure.defaults`. @billsacks can you please confirm this? If so I will change this line to:
```
ARCH_LOCAL      =       -DNEC -DNONSTANDARD_SYSTEM_SUBR -DWRF_USE_CLM  CONFIGURE_D_CTSM
```

davegill replied:
@negin513 
Negin,
Yes, I agree that everywhere you have replaced `-DWRF_USE_CLM` with `CONFIGURE_D_CTSM`, we need to have BOTH of those options listed.

billsacks replied:
@davegill - This was an intentional change that I made. The problem with having `-DWRF_USE_CLM` when we're compiling with CTSM is that there are module name collisions at link time, because many of the modules in module_sf_clm have the same name as modules in the CTSM library. So the logic here is to turn off `-DWRF_USE_CLM` if the user is building with CTSM, but to keep it on otherwise. We have a partially-complete project to rename all modules in CTSM to add a `ctsm_` prefix, which would get around this issue. Once that's done, we could have both `-DWRF_USE_CLM` and `-DWRF_USE_CTSM` at the same time, but I'm not sure what the priority is of that. If you feel it's important to keep the old CLM code compiled in the same build as CTSM, then we can bump up the priority of that rename to allow this.

negin513 replied:
@davegill based on what @billsacks is saying, I should go ahead and remove the `-DWRF_USE_CTSM` flag. or @davegill do you think some other solution that takes care of this?

davegill commented:
@negin513 
Negin,
Please add your new doc/README.CTSM to the list of files in your PR. The level of detail contained is spot on.

negin513 commented:
Hello Dave,
I made the following changes regarding your comments.

> 1. Would you add a doc/README.CTSM file. In can be short. Info on where to go for web resources, etc. A few sentences.

I've added a file called` doc/README.CTSM` file and included a few paragraphs in it. :-)

> 2. Do you have an easy example that you can show using WRF+CTSM vs WRF with any other land model? A couple of figures to demonstrate that things seem to be working OK. That can go in the "test conducted" section.

I've added an image including Bias for Tmin and Tmax for April 2013 comparing with NOAH, NOAH-MP, and observation. We do have a much more in-depth analysis of the simulations for a paper that is under-preparation.

negin513 commented:
> @negin513
> Negin,
> Please add your new doc/README.CTSM to the list of files in your PR. The level of detail contained is spot on.

Done :-)

davegill commented:
@barlage 
Mike,
Do we have your affiliation correct?

negin513 commented:
> @barlage
> Mike,
> Do we have your affiliation correct?

I've updated the affiliations.

davegill commented:
@negin513 
Negin,
I think that we can find an elegant solution to the cpp flags. Hold off on the changes to that file. Literally, we have MONTHS before we are in a crunch for time. We can probably get something that will work figured out by tomorrow.

davegill commented:
@weiwangncar @dudhia 
Folks,
The NCAR CTSM group would like to merge their capability onto the develop branch. 
1. The jenkins tests all pass, so there do not seem to be any side effects when NOT using their scheme.
2. The PR has figures that demonstrate the impact of utilizing the scheme.

What else is required to get this code rolled into WRF?

dudhia commented:
I will take a look and see if I have any questions.

On Thu, Aug 20, 2020 at 4:36 PM Dave Gill <notifications@github.com> wrote:

> @weiwangncar <https://github.com/weiwangncar> @dudhia
> <https://github.com/dudhia>
> Folks,
> The NCAR CTSM group would like to merge their capability onto the develop
> branch.
>
>    1. The jenkins tests all pass, so there do not seem to be any side
>    effects when NOT using their scheme.
>    2. The PR has figures that demonstrate the impact of utilizing the
>    scheme.
>
> What else is required to get this code rolled into WRF?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/wrf-model/WRF/pull/1275#issuecomment-677940865>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEIZ77B4GQWLYIL3BJP4RL3SBWQNRANCNFSM4QEGIXJA>
> .
>

weiwangncar commented:
@davegill Can you explain how 'feature branch' work? Will the code remain on the feature branch after the next major release?

davegill commented:
@weiwangncar 
Wei,
The merge went SMOOTHLY. There is actually no need for the feature branch. This is ready to go onto develop. The code changes have ifdefs where required. The jenkins testing is OK.

weiwangncar commented:
Ok, I will also take a look at the code more closely.

weiwangncar commented on README.md:
```diff
@@ -1,5 +1,9 @@
 ### WRF-ARW Modeling System  ###
 
+This is the WRF feature branch for coupling WRF with CTSM [The Community Terrestrial Systems Model](https://github.com/ESCOMP/CTSM).
```
If this is going to merge onto the develop branch, then it is no longer a feature branch. Can you make a change here? Perhaps the change in this file is no longer necessary?

negin513 replied:
You are right. I've just removed these changes in README.md.

weiwangncar commented on doc/README.CTSM:
```diff
@@ -0,0 +1,25 @@
+For comprehensive instructions on how to run WRF with CTSM please check: 
+https://escomp.github.io/ctsm-docs/versions/master/html/lilac/specific-atm-models/wrf.html
+
+In summary, he procedure for running WRF-CTSM is similar to the workflow for 
```
Typo here: it should be 'the', not 'he'.
Also, perhaps the info on CTSM can be added here (e.g. the GitHub location).

negin513 replied:
Thanks for your comment. I just corrected the typo.

weiwangncar commented on phys/module_surface_driver.F:
```diff
@@ -3863,6 +3869,212 @@ SUBROUTINE surface_driver(                                         &
+
+#else
+        CALL wrf_error_fatal( "SF CTSM CURRENTLY ifdef'ed out, set -DWRF_USE_CTSM by setting&
+             & WRF_CTSM_MKFILE before running configure" )
```
Should this be part of module_check_a_mundo.F?

negin513 replied:
You are correct! This is not needed here, as a similar check is done in `module_check_a_mundo.F`.  I removed this conditional from phy/module_surface_driver.F. @BillSacks do you agree with this change?

billsacks replied:
When I originally put this in (in 4e9f24e61), my feeling was that it doesn't hurt to have redundant error checks like this – and in many cases it can really help, both to catch issues as the code evolves in the future, and to make the expected behavior clear to readers. However, it looks like 2296313b introduced a bug in this logic by moving where the ifdef occurs relative to the case statement. So, as currently written, it looks like the `#else` is associated with the previous case (for CLMSCHEME), which (if I'm reading the code right) would lead to an abort whenever someone tries to run with CLMSCHEME! It looks like the same issue was introduced in `module_physics_init.F` in that commit.

So my feeling is:

- Best option: back out the changes to module_physics_init.F and module_surface_driver.F from 2296313b (unless those changes are needed for some reason?) This will mean a redundant error check, but again, my feeling is it's better to err on the side of redundancy in error checks in cases like this.

- Okay option: remove the `#else` block from both here and module_physics_init.F.

negin513 replied:
> However, it looks like [2296313](https://github.com/wrf-model/WRF/commit/2296313bf3ff90166df5063548d05f8b1cdfe80b) introduced a bug in this logic by moving where the ifdef occurs relative to the case statement. So, as currently written, it looks like the `#else` is associated with the previous case (for CLMSCHEME), which (if I'm reading the code right) would lead to an abort whenever someone tries to run with CLMSCHEME!

Bill, I am not sure if I see correctly how does this `else` is triggered when CLMSCHEME is used. I believe this `endif` https://github.com/wrf-model/WRF/blob/9f663273504bf89e54ead7757ac00290e293520e/phys/module_surface_driver.F#L3870 is related to the CLMSCHEME 
Am I missing something?

negin513 replied:
Anyway, I am going to remove the else block as Wei suggested. I believe this is redundant and this code block will never get triggered because this check happens in `module_check_a_mundo.F`.

https://github.com/wrf-model/WRF/blob/9f663273504bf89e54ead7757ac00290e293520e/phys/module_surface_driver.F#L4073-L4076

negin513 replied:
Thanks, @billsacks to pointing me out to this redundant error check in`module_physics_init.F`. I removed it from there as well.

billsacks replied:
> Bill, I am not sure if I see correctly how does this `else` is triggered when CLMSCHEME is used.

With the changes you have made, this doesn't matter any more, but to try to bring clarity to this discussion: Rereading my earlier message, I realized that I didn't state it very clearly. It's true that the preprocessor block related to CLM was ended with the earlier `#endif`. But the problem was that the `#else` block associated with WRF_USE_CTSM would have appeared under the CASE block for CLM. (I'm still not sure if I'm explaining this clearly.)

weiwangncar commented on phys/module_surface_driver.F:
```diff
@@ -3863,6 +3869,212 @@ SUBROUTINE surface_driver(                                         &
+         PRESENT (smoiseq)        .AND.                              &
+         PRESENT (wtddt)          .AND.                              &
+         PRESENT (stepwtd)        .AND.                              &
+#endif                   
```
What's the reason to check the presence of these variables?

negin513 replied:
@weiwangncar , you are absolutely right. Checking if these variables are present is not necessary at all. I copied this from NOAH-MP, as I thought we might want to use similar to NOAH-MP handling of groundwater in here. But we don't need this here at all. I removed them.

weiwangncar commented:
Perhaps something regarding the removal of CLM compilation when using CTSM option should be added to README.CTSM. One can imagine users may think of comparing results with these two options, and not knowing CLM isn't compiled when CTSM is used.

dudhia commented on arch/configure.defaults:
```diff
@@ -16,7 +16,7 @@ CC              =       CONFIGURE_CC
 RWORDSIZE       =       CONFIGURE_RWORDSIZE
 PROMOTION       =       #-fdefault-real-8
-ARCH_LOCAL      =       -DNEC -DNONSTANDARD_SYSTEM_SUBR  -DWRF_USE_CLM
+ARCH_LOCAL      =       -DNEC -DNONSTANDARD_SYSTEM_SUBR CONFIGURE_D_CTSM
```
I was wondering why -DWRF_USE_CLM was replaced. This is the flag to disable CLM compilation that we may still want. It might have been added because CLM took a long time to compile on some platforms.

dudhia replied:
I guess the question is whether CTSM is intended to completely replace CLM.

davegill replied:
@dudhia 
Jimy,
The flag `CONFIGURE_D_CTSM` is now used for both. In the Config.pl script, if the LILAC/CTSM build is requested, then a flag for CTSM is supplied. If the standard build is requested, then the flag -DWRF_USE_CLM is supplied. The LILAC/CTSM and CLM schemes cannot co-exist due to namespace collision. This is a functional and correct implementation that allows both capabilities, but selected at compile-time.

dudhia commented:
OK, so this choice occurs as an option in the configure stage? How/when
does it appear?

On Tue, Aug 25, 2020 at 9:56 AM Dave Gill <notifications@github.com> wrote:

> *@davegill* commented on this pull request.
> ------------------------------
>
> In arch/configure.defaults
> <https://github.com/wrf-model/WRF/pull/1275#discussion_r476557539>:
>
> > @@ -16,7 +16,7 @@ CC              =       CONFIGURE_CC
>  LD              =       $(FC)
>  RWORDSIZE       =       CONFIGURE_RWORDSIZE
>  PROMOTION       =       #-fdefault-real-8
> -ARCH_LOCAL      =       -DNEC -DNONSTANDARD_SYSTEM_SUBR  -DWRF_USE_CLM
> +ARCH_LOCAL      =       -DNEC -DNONSTANDARD_SYSTEM_SUBR CONFIGURE_D_CTSM
>
> @dudhia <https://github.com/dudhia>
> Jimy,
> The flag CONFIGURE_D_CTSM is now used for both. In the Config.pl script,
> if the LILAC/CTSM build is requested, then a flag for CTSM is supplied. If
> the standard build is requested, then the flag -DWRF_USE_CLM is supplied.
> The LILAC/CTSM and CLM schemes cannot co-exist due to namespace collision.
> This is a functional and correct implementation that allows both
> capabilities, but selected at compile-time.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/wrf-model/WRF/pull/1275#discussion_r476557539>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEIZ77ALJBXIGNJAPT3OVDTSCPNJ5ANCNFSM4QEGIXJA>
> .
>

davegill commented on doc/README.CTSM:
```diff
@@ -0,0 +1,25 @@
+
+1. Clone the CTSM repository,
+2. Build CTSM libraries and LILAC,
+3. Define namelist options reuired for CTSM,
```
"required"

negin513 replied:
Typo fixed. Thanks. :-)

davegill commented:
@negin513 @dudhia 
> OK, so this choice [to choose CLM vs CTSM] occurs as an option in the configure stage? How/when does it appear?

Negin,
Would you address this please?

dudhia commented:
I see that there are detailed instructions which include environment
settings.

On Tue, Aug 25, 2020 at 10:07 AM Dave Gill <notifications@github.com> wrote:

> @negin513 <https://github.com/negin513> @dudhia
> <https://github.com/dudhia>
>
> OK, so this choice [to choose CLM vs CTSM] occurs as an option in the
> configure stage? How/when does it appear?
>
> Negin,
> Would you address this please?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/wrf-model/WRF/pull/1275#issuecomment-680120091>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEIZ77BEWAZMTCBWFJ3D35TSCPOTLANCNFSM4QEGIXJA>
> .
>

negin513 commented:
> Perhaps something regarding the removal of CLM compilation when using CTSM option should be added to README.CTSM. One can imagine users may think of comparing results with these two options, and not knowing CLM isn't compiled when CTSM is used.

@weiwangncar I've added a sentence in README.CTSM pointing that if the CTSM is being used for compiling, the user cannot use CLM option (sf_surface_physics=5) in their namelist later.

negin513 commented:
> I see that there are detailed instructions which include environment settings.



Thanks @dudhia for reviewing this PR. You are correct. This option is set by explicitly setting the environment variable `WRF_CTSM_MKFILE`. The detailed instructions are [here](https://escomp.github.io/ctsm-docs/versions/master/html/lilac/specific-atm-models/wrf.html#building-wrf-with-ctsm).
```
setenv WRF_CTSM_MKFILE /glade/scratch/$USER/WRF-CTSM/CTSM/ctsm_build_dir/bld/ctsm.mk
```

dudhia commented:
Without CTSM do we still have the option of not compiling CLM as we did
before with the -DWRF_USE_CLM.

On Wed, Aug 26, 2020 at 6:25 PM Negin Sobhani <notifications@github.com>
wrote:

> *@negin513* commented on this pull request.
> ------------------------------
>
> In doc/README.CTSM
> <https://github.com/wrf-model/WRF/pull/1275#discussion_r477765806>:
>
> > +  Using CTSM with WRF
> +=====================
> +
> +CTSM has now been coupled to WRF.
> +CTSM can be used as the land surface model scheme in WRF by choosing sf_surface_physics = 6
> +in the WRF namelist.input file.
> +
> +For comprehensive instructions on how to run WRF with CTSM please check:
> +https://escomp.github.io/ctsm-docs/versions/master/html/lilac/specific-atm-models/wrf.html
> +
> +In summary, he procedure for running WRF-CTSM is similar to the workflow for
> +running WRF real cases, except that it requires additional steps to:
> +
> +1. Clone the CTSM repository,
> +2. Build CTSM libraries and LILAC,
> +3. Define namelist options reuired for CTSM,
>
> Typo fixed. Thanks. :-)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/wrf-model/WRF/pull/1275#discussion_r477765806>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEIZ77HIG5RRTA54ZSMIOZLSCWRX5ANCNFSM4QEGIXJA>
> .
>

davegill commented:
@dudhia 
> Without CTSM do we still have the option of not compiling CLM as we did before with the -DWRF_USE_CLM.

Jimy,
Without CTSM being built, the procedure is identical. A user edits the configure file and changes the `-DWRF_USE_CLM` flag. Either the flag is removed, or the name could be changed, such as `-DWRF_NO_USE_CLM`. This is exactly what users do today to get rid of the CLM build.

dudhia commented:
Thanks.

On Thu, Aug 27, 2020 at 9:16 AM Dave Gill <notifications@github.com> wrote:

> @dudhia <https://github.com/dudhia>
>
> Without CTSM do we still have the option of not compiling CLM as we did
> before with the -DWRF_USE_CLM.
>
> Jimy,
> Without CTSM being built, the procedure is identical. A user edits the
> configure file and changes the -DWRF_USE_CLM flag. Either the flag is
> removed, or the name could be changed, such as -DWRF_NO_USE_CLM. This is
> exactly what users do today to get rid of the CLM build.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/wrf-model/WRF/pull/1275#issuecomment-682014429>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEIZ77DR7ZTXPJBGNCLWF5LSCZ2GXANCNFSM4QEGIXJA>
> .
>

dudhia commented on phys/module_ra_cam.F:
```diff
@@ -907,7 +908,8 @@ subroutine camrad(RTHRATENLW,RTHRATENSW,                           &
       enddo
       endif
-  CASE (CLMSCHEME)
+
```
as far as I can tell, no change is needed in module_ra_cam.F

negin513 replied:
Thanks for the comment. 
As far as I can tell, no change has been done for this file. 
It seems like @billsacks previously added some todos here which is removed in 83dbf7b6, but caused the extra line we see here. I actually revert this back to be identical to before. So basically, this file should not even show up in the files changed list.

davegill commented on share/module_check_a_mundo.F:
```diff
@@ -2107,6 +2107,31 @@ END FUNCTION bep_bem_nbui_max
+         CALL wrf_debug ( 0, TRIM(wrf_err_message) )
+         wrf_err_message = '--- ERROR: However, the WRF CTSM scheme was not compiled in WRF.'
+         CALL wrf_debug ( 0, TRIM(wrf_err_message) )
+         wrf_err_message = '--- ERROR: Please set -DWRF_USE_CTSM by setting WRF_CTSM_MKFILE&
```
Perhaps send the user to a README file? Instead of the "Please set ...", the info in the already-generated README would be better. The first couple of lines are good to keep.

negin513 replied:
100% agreed, I don't like the message  "Please set -DWRF_USE_CTSM ...".
I changed it to "Please read doc/README.CTSM for how to compile WRF with CTSM." as you suggested.

weiwangncar commented on dyn_em/solve_em.F:
```diff
@@ -128,6 +128,8 @@ SUBROUTINE solve_em ( grid , config_flags  &
    LOGICAL                        :: diag_flag
    INTEGER                        :: ke_diag ! tells reflectivity calculation whether to do full depth or only k=1
+   LOGICAL                        :: restart_flag
+   LOGICAL                        :: restart_run
```
@negin513 Is restart_run used anywhere?

negin513 replied:
Thanks @weiwangncar . Correct! `restart_run` variable was previously used here but  now it is not being used here at all.

negin513 replied:
I remove the unused variable.

weiwangncar commented:
It looks like the things I pointed out are all addressed (except the one posted today).

weiwangncar commented:
I'm now ok with this PR. Thanks.

davegill commented on arch/preamble:
```diff
@@ -100,6 +100,9 @@ NETCDF4_DEP_LIB = $(DEP_LIB_PATH) $(HDF5) $(ZLIB) $(GPFS) $(CURL)
 
+#### CTSM pieces
+
+# CTSMINCLUDEGOESHERE
```
@negin513 
Negin,
Is this used?

billsacks replied:
Yes, this is used. I followed some other examples in this file. It is used by this line in Config.pl:

```perl
$_ =~ s:\# CTSMINCLUDEGOESHERE:include $sw_ctsm_mkfile_path: ;
```

davegill replied:
Thanks

davegill commented on dyn_em/module_first_rk_step_part1.F:
```diff
@@ -579,6 +581,7 @@ SUBROUTINE first_rk_step_part1 (   grid , config_flags              &
      &        ,LakeModel=grid%sf_lake_physics,  lake_min_elev=grid%lake_min_elev    & !lake  
 #if ( EM_CORE == 1)
      &        ,LakeMask=grid%LakeMask                                               & ! lake
+     &        ,restart_flag=restart_flag                                               & ! is it a restart timestep?
```
@negin513 
Negin,
Line up the trailing "&" + comment indenting/spacing. This is probably easiest from a sandbox, as I can never guess space vs tab, etc

negin513 replied:
Thanks for your comment. I've addressed your comment in the recent push.

davegill commented on dyn_em/solve_em.F:
```diff
@@ -293,6 +295,13 @@ SUBROUTINE solve_em ( grid , config_flags  &
+! Set restart flag value history output time
+!-----------------------------------------------------------------------------
+!
+      restart_flag = .false.
```
@negin513 
Negin,
Line up the top restart_flag with the following IF test

negin513 replied:
Thanks for the suggestion. I made that correction.

davegill commented on phys/module_surface_driver.F:
```diff
@@ -3861,6 +3867,186 @@ SUBROUTINE surface_driver(                                         &
+#ifdef WRF_USE_CTSM
+     CASE (CTSMSCHEME)
+
+! TODO (NS): Adding errors and warning on compatible PBL schemes, etc. For
```
@negin513 
Negin,
Has this been moved into check_a_mundo? This "TODO" should be removed.

negin513 replied:
@davegill No. This has not been moved to check_a_mundo. I put this here similar to what was there for other land models for example: 
https://github.com/wrf-model/WRF/blob/94df8047e63c91e147c1c7ce803485e2d7656e51/phys/module_surface_driver.F#L3631-L3635

negin513 replied:
I just removed the TODO comment from this file.

davegill replied:
Negin,
All good - I just wanted that "TODO" thing removed.

davegill commented:
@negin513 
Negin,
The previous pieces are a last sweep through for clean-up purposes. 

The PR commit message lists 18 files, and github says 16 files are changed.

davegill commented:
@negin513 
Negin,
Last thing: The PR commit message lists 17 files, and github says 16 files are changed.

negin513 commented:
> @negin513
> Negin,
> The previous pieces are a last sweep through for clean-up purposes.
> 
> The PR commit message lists 18 files, and github says 16 files are changed.

@davegill Thanks so much for reviewing this. :-)

The PR first message listed README and phys/module_ra_cam.F as changed files but the changes in those two files are reverted. I updated the PR message to show the 16 changed files:

M Registry/Registry.EM_COMMON
M arch/Config.pl
M arch/configure.defaults
M arch/postamble
M arch/preamble
A doc/README.CTSM
M dyn_em/module_first_rk_step_part1.F
M dyn_em/module_initialize_real.F
M dyn_em/solve_em.F
M main/depend.common
M phys/Makefile
M phys/module_physics_init.F
A phys/module_sf_ctsm.F
M phys/module_surface_driver.F
M share/input_wrf.F
M share/module_check_a_mundo.F

davegill commented:
@weiwangncar @dudhia 
Folks,
I am OK with this PR. I am approving it (from a software / PR conformity perspective). I would like one of the physics team to also approve from the perspective of the merits of the scheme.

weiwangncar commented:
@davegill Should we inform the DA group so that they can conduct a compilation test?

davegill commented:
@liujake 
Jake,
Would you please conduct some simple tests to see if this brakes anything related to DA. The forward model group is happy.

vikramjakhr commented:
reopen testing

vikramjakhr commented:
reopen

vikramjakhr commented:
reopen

davegill commented:
@negin513 @billsacks 
Folks,
The CTSM capability is in the WRF develop branch.

negin513 commented:
> @negin513 @billsacks 
> 
> Folks,
> 
> The CTSM capability is in the WRF develop branch.

Thanks so much for your help with this.
